Fixes #31781 - Update existing API data from the API middleware

amirfefer opened the pull request:
Redux store should be one source of truth and in some cases, an update is needed (in order to avoid data duplication).
This PR is used in https://github.com/theforeman/foreman/pull/8285

amirfefer committed:
Fixes #31781 - Update existing API data from the API middleware

ezr-ondrej reviewed (commented):
I don't quite get why are we using `UPDATE`, first time I've read it, I though it's gonna make some mutations on the server.
This feels bit weird, it feels like we can manipulate the data from React however we want, but it shouldn't be the case, right?

I feel that this puts a lot of burden on the user of the `UPDATE` action, as he need to know exactly what to update it with (to know the exact structure of updated data).

Wouldn't we be better of to reload such chunk of data after a modification to it? It feels much more error prone and it shouln't be a big issue as it only adds one API request after the mutation has happend.

Let me know what you think :)

amirfefer commented:
The API store is our data "warehouse" for resources given from ajax calls, before the API middleware we had a reducer for each resource, including the option to alter the current raw data, but the API middleware causes a regression due to the lack of the option to alter the data, hence the data is forced to be read-only from the client side, and can be only be updated after an ajax call. this situation forces us to make two API calls on every change (one for delete or update a specific resource, the second is to get all the resources list for keeping it up to date).

The other option is to copy the data in another store key, but this requires boilerplate (creating new reducers, new actions) and removes  the store `one source of truth` important behavior (i.e two components that dependant on the same resources, which part of the store is the right one? )

It's important to notice that we should not alter the data in selectors nor outside reducers, selectors are made mainly for transforming raw data into presentable data (i.e sorting) and for performances issues (memorization)

ezr-ondrej commented:
In a disscussion we've decided to offer simple action, that will perform an server side mutation and update the data in store with the response data from given action and that should be the aim of this PR.
I've shared the disscusion for better visibility on discourse: https://community.theforeman.org/t/react-vs-api-interactions-improvements/22326 if anyone has some major idea, please share it there :)

amirfefer commented:
Thanks @ezr-ondrej ,

I updated the PR, the UPDATE action is part of the APIMiddleware, consumers can update the corresponding resources list just after a successful API call:

```js
// This action creates a post API call and only if it ends successfully,
// the response object will update with the given 'updatedData' object

dispatch(
  post({
    key: MY_SPECIAL_KEY,
    url,
    updatedData: {results: ['altered resource']}
  })
);
```

ezr-ondrej commented:
Does this make the actual API update request, or it needs to be triggered manually?
Can't we have an `updateResource(resource_id, update_url, updated_attributes_object)` kind of action, that would:
1. trigger the update with the updated data
2. updated the redux data with API response from such action

jeremylenz commented:
The PR title "Make Redux store overridable" is a little misleading IMO.  I think we are using Redux as intended-- we're just adding an action that, just like any other Redux action, affects the data using reducers.  We're not 'mutating the Redux store' or overriding it or doing anything really unorthodox.

jeremylenz commented on webpack/stories/docs/api-middleware.stories.mdx:
```diff
@@ -150,3 +150,23 @@ dispatch(
+
+## Update resources object
+
+There are some cases in which the resources list, that fetched earlier, needs to be altered (update, delete), instead of making two separated API calls (for altering and fetching the up-to-date resources list),
```
```suggestion
After an API request SUCCESS action, it is often reconfirmed with a second GET request to view the updated data, and keep the Redux store up to date.  However, if we receive a SUCCESS it is acceptable to assume that the data was actually updated as requested, and not completely necessary to make the additional request.  To facilitate this,
```

jeremylenz commented on webpack/stories/docs/api-middleware.stories.mdx:
```diff
@@ -150,3 +150,23 @@ dispatch(
+## Update resources object
+
+There are some cases in which the resources list, that fetched earlier, needs to be altered (update, delete), instead of making two separated API calls (for altering and fetching the up-to-date resources list),
+the API middleware can accept an `updatedData` object, with the updated resource object corresponding to its key. 
```
```suggestion
the API middleware can accept an `updatedData` object, with the updated resource object corresponding to its key. 
Thus, Redux's copy of the backend data is changed based on this trust, and an additional API request to show the updated data is avoided.
```

jeremylenz commented on webpack/assets/javascripts/react_app/redux/API/APIRequest.js:
```diff
@@ -46,6 +50,13 @@ export const apiRequest = async (
       response: response.data,
     });
 
+    updatedData &&
```
I like the fact that this UPDATE action is triggered from within the API middleware.  If the action were more easily available, there would be a bigger risk of the Redux store and backend becoming out of sync.  But here, it provides an opinionated workflow and should avoid a lot of potential bugs. :+1:

jeremylenz commented on webpack/stories/docs/api-middleware.stories.mdx:
```diff
@@ -150,3 +150,23 @@ dispatch(
+  post({
+    key: MY_SPECIAL_KEY,
+    url,
+    updatedData: {results: ['altered resource']}
```
What format must this data be in?  I'm guessing `updatedData` wants a data shape identical to the entire response?  Maybe it's good to call that out above.

amirfefer replied:
not the all response object, but only the `data` part of it,
it  depends on the data structure of each API call, mostly it is something like 
```js
{ results: [{id: 1 , name: 'A'}, {id: 2 , name: 'B'}] } 
```
But this structure might be different.

jeremylenz reviewed (commented):
Looking good!  Thanks @amirfefer and @ezr-ondrej for driving the discussion this morning :)

suggestions & comments below.

amirfefer commented:
Thanks @jeremylenz and @ezr-ondrej for the review!

> The PR title "Make Redux store overridable" is a little misleading IMO. I think we are using Redux as intended-- we're just adding an action that, just like any other Redux action, affects the data using reducers. We're not 'mutating the Redux store' or overriding it or doing anything really unorthodox. (@jeremylenz )

Yes, I Agree, how about  something like " Update existing API data directly from the API middleware"

> Does this make the actual API update request, or it needs to be triggered manually?

This makes the actual update API, and if it ends with success, the store is updated with the given data.
 
> Can't we have an updateResource(resource_id, update_url, updated_attributes_object) kind of action, that would:
trigger the update with the updated data
updated the redux data with API response from such action (@ezr-ondrej )

This addition does it both with one dispatch but from the API middleware, it reduces boilerplate and potential bugs.  dispatching something like this
```js
dispatch(
  post({
    key: MY_SPECIAL_KEY,
    url,
    updatedData: {results: ['altered resource']}
  })
);
```
 creates a post and also updates the redux store only if the API ended successfully.

jeremylenz commented:
> Yes, I Agree, how about something like " Update existing API data directly from the API middleware"

that sounds good to me! :+1:

ezr-ondrej commented on webpack/stories/docs/api-middleware.stories.mdx:
```diff
@@ -150,3 +150,26 @@ dispatch(
+  post({
+    key: MY_SPECIAL_KEY,
+    url,
+    updatedData: {results: ['altered resource']}
```
Why do we still send the data to the UPDATE action? Can't we get the updated data from the API response?

amirfefer replied:
AFAIK post requests (which alter a state) don't have to return the updated resource (i.e forget_status action) so I left it controlled by the consumer.

ezr-ondrej replied:
no, they don't need to, but they usually do. I was talking about it in the meeting, we should state that returning the altered resource is a standard and only if we have a real case where we don't we would allow manual updates, but the updatedData seems very wrong.
If you insist on allowing this, I strongly believe you shouldn't be able to update all the data through that, but only the data that got updated, so it should be pair of `id` and `data` where data is data only for the given id.

Ron-Lavi commented:
Hey @amirfefer what's the status here? :) can you rebase this PR to refresh the code a little bit?

amirfefer commented:
@laviro , rebased :)

Ron-Lavi commented on webpack/assets/javascripts/react_app/redux/API/APIRequest.js:
```diff
@@ -46,6 +50,13 @@ export const apiRequest = async (
+      dispatch({
+        type: UPDATE,
+        key,
+        payload: updatedData,
```
Can we pass here the `prevState` into the `updateData` as a callback?

the consumer can use it similiarly to React `useState(prevState => ({ ...prevState, ...changedState}))`

so I guess the consumer can do something like:
```js
post({
    key: MY_SPECIAL_KEY,
    url,
    params: { id }
    updateData: prevState => ({
        ...prevState,
        results: prevState.results.map(res => res.id === id ? changedResults : res)})
})
```

I know that you used `deep` update to redux store, but maybe as a callback user have more power and understanding of how this works.

ezr-ondrej replied:
this is indeed better, then just requiring the full updated data blob, but I still believe we should be able to update the store from the response without the need of updateData.

Ron-Lavi replied:
@ezr-ondrej if I understand correctly you mean triggering a `post(key, url, params)`, and right after that an `update(key, payload)` ?

amirfefer replied:
👍 for this callback idea @laviro
@ezr-ondrej, the `updateData` callback is optional, if the consumer won't add it to the api request, the response will be updated in the store according to the corresponding key:
```js
    const response = await getApiResponse({ type, url, headers, params });

    dispatch({
      type: SUCCESS,
      key,
      payload: modifiedPayload,
      response: response.data, // adds the data to the store
                              //  if the API returns the altered resources,
                             //  no need to pass the updatedData callback
    });

// optional by the consumer, should be used when api requests dosen't return all the altered resources
    updatedData && 
      dispatch({
        type: UPDATE,
        key,
        payload: updatedData(getState().key.response), // updatedData is a callback with prevState as argument
      });
```

This modification is for keeping the store once source of truth without duplicate the data

Ron-Lavi reviewed (commented):
Overall looks awesome! I would suggest using it as a callback to bring more control of the state update to the consumer, e.g: 
```js
post({
    key: MY_SPECIAL_KEY,
    url,
    params: { id }
    updateData: prevState => { /* adjust old state here **/ }
```

amirfefer commented:
Thanks, @laviro, @ezr-ondrej 

- [x] rebased
- [x] switched to a callback driven approch 
- [x] updated tests
- [x] updated docs

`updatedData` callback accepts two arguments, prevState and currentResponse:

#### prevState
The previous response data which stored on redux store and corresponding with the given key.

#### currentResponse
The current API call response.


Example:

```js
// This action creates a post API call and only if it ends succesfully,
// the response object will update with the returned data from 'updatedData' callback 
dispatch(
  post({
    key: MY_SPECIAL_KEY,
    url,
    updatedData: (prevResponseState, currentResponse) => ({
      results: prevStateData.filter(
        item => item.id !== currentResponse.data.id
      ),
    }),
  })
);
```

Ron-Lavi commented on webpack/assets/javascripts/react_app/redux/API/APIRequest.js:
```diff
@@ -46,6 +52,13 @@ export const apiRequest = async (
       response: response.data,
     });
 
+    updatedData &&
```
I think it should be called `updateData` as it's a function now

Ron-Lavi commented on webpack/assets/javascripts/react_app/redux/API/APIFixtures.js:
```diff
@@ -26,6 +27,20 @@ export const action = {
+export const postActionWithCallback = {
+  type: API_OPERATIONS.POST,
+  payload: {
+    key: 'PREV_STATE',
```
This key might be a bit confusing, how about `MY_RESOURCE`

amirfefer replied:
agree, though the reason for the `PREV` prefix is to emphasize  the **old** state (before fetching the data), I've changed it to ` INITIAL_RESOURCE`

Ron-Lavi commented on webpack/stories/docs/api-middleware.stories.mdx:
```diff
@@ -150,3 +150,37 @@ dispatch(
+    url,
+    updatedData: (prevResponseState, currentResponse) => ({
+      results: prevStateData.filter(
+        item => item.id !== currentResponse.data.id
```
I think in the implementation you already pass currentResponse.data,
So here it should be:

```suggestion
        item => item.id !== currentResponse.id
```

Ron-Lavi reviewed (commented):
Nice work, left a few comments,
Btw, is there a corresponding PR that uses it? Thanks @amirfefer

amirfefer commented:
Thanks @laviro
yes, this [PR](https://github.com/theforeman/foreman/pull/8285/files#diff-66ceed3910b85c999c7853c408ec031c8a2f699574cb62878939a1646a5a442bR13-R22) uses this feature.

Ron-Lavi commented on webpack/assets/javascripts/react_app/redux/API/APIReducer.js:
```diff
@@ -36,6 +36,15 @@ const apiReducer = (state = initialState, { type, key, payload, response }) => {
+            response: payload,
+          },
+        },
+        { deep: true }
```
should we still use `deep` now that we pass the `prevState` and expecting a modified State?

amirfefer replied:
yes, we don't need this "insurance"  anymore

Ron-Lavi commented on webpack/assets/javascripts/react_app/redux/API/APIRequest.js:
```diff
@@ -46,6 +52,13 @@ export const apiRequest = async (
       response: response.data,
     });
 
+    updateData &&
```
Can you also check if the state for the API key exists?
e.g.:
```suggestion
   const prevResponse = selectAPIResponse(prevState, key);
   if (updateData && !isEmpty(prevResponse)) dispatch....
```

another option would be to display show an error in the console.

amirfefer replied:
The state can be empty, consumers can modify just the current response's data,
I don't think that raising an error is needed here, if a consumer will use an unexisting state it would be just empty object (and trying to access undefined keys will end with an error).

```js
  post({
    key: MY_SPECIAL_KEY,
    url,
    updateData: (prevResponseState, currentResponse) => ({
      results: currentResponse.map(someCallback),
    }),
```

Mostly in this case it would be better to change the return format in the server, though we don't change the API often (the API serves other usages) therefore there are some edge cases for this ability IMHO.

Ron-Lavi replied:
rethinking about it, the namespace will always be there as it's being dispatched after the REQUEST stage, 
so no concern here

Ron-Lavi commented on webpack/stories/docs/api-middleware.stories.mdx:
```diff
@@ -150,3 +150,37 @@ dispatch(
+// This action creates a post API call and only if it ends succesfully,
+// the response object will update with the returned data from 'updateData' callback 
+dispatch(
+  post({
+    key: MY_SPECIAL_KEY,
+    url,
+    updateData: (prevResponseState, currentResponse) => ({
+      results: prevStateData.filter(
+        item => item.id !== currentResponse.id
+      ),
+    }),
```
I guess that most of the times the server won't return anything in POST requests,
so might be valuable to show something like:

```js
  post({
    key: MY_SPECIAL_KEY,
    url,
    params: { id: selectedID } // selectedID can be passed to the action
    updateData: (prevResponseState, currentResponse) => ({
      results: prevStateData.filter(
        item => item.id !== selectedID
      ),
    }),
```

amirfefer replied:
post requests usually do return some data, in this example there's no use for the `currentResponse` though.
I just created a simple example that shows a usage for both of the arguments.

The more the better :) I've added this example

Ron-Lavi reviewed (commented):
Almost there @amirfefer! big thanks

Ron-Lavi reviewed (approved):
Thanks @amirfefer!! code-wise looks great,
didn't test it yet with the other PR though

ezr-ondrej reviewed (approved):
I'd love if we would have a default processing, because this is usually update single resource and we should be handle it by default.
But I agree this is much better then what we have now :+1:

ezr-ondrej commented:
Thanks @laviro and @jeremylenz for the reviews!
